internal/resolver/dns/dns_resolver.go

```diff
@@ -201,28 +204,38 @@ func (d *dnsResolver) Close() {
 
 func (d *dnsResolver) watcher() {
 	defer d.wg.Done()
+	backoffIndex := 1
 	for {
-		select {
-		case <-d.ctx.Done():
-			return
-		case <-d.rn:
-		}
-
 		state, err := d.lookup()
 		if err != nil {
+			// Report error to the underlying grpc.ClientConn.
 			d.cc.ReportError(err)
 		} else {
-			d.cc.UpdateState(*state)
+			err = d.cc.UpdateState(*state)
 		}
 
-		// Sleep to prevent excessive re-resolutions. Incoming resolution requests
-		// will be queued in d.rn.
-		t := time.NewTimer(minDNSResRate)
+		var timer *time.Timer
+		if err == nil {
+			// Success resolving, wait for the next ResolveNow. However, also wait 30 seconds at the very least
+			// to prevent constantly re-resolving.
+			backoffIndex = 1
+			timer = time.NewTimer(minDNSResRate)
```

Comment:
IMO this should use `newTimer`, too, or else we can't test all the timing aspects that we _should_ be testing.

Reply:
newTimer is extremely coupled to like 15 tests. The success path (with MinDNSResRate) is already tested within the unit tests. Ex: https://github.com/zasweq/grpc-go/blob/move-backoff-to-dns-resolver/internal/resolver/dns/dns_resolver_test.go#L1046.